examples/applications/plot_cyclical_feature_engineering.py

```diff
@@ -0,0 +1,693 @@
+    cyclic_cossin_transformer,
+    RidgeCV(alphas=alphas),
+)
+evaluate(cyclic_cossin_linear_pipeline, X, y, cv=ts_cv)
+
+
+# %%
+#
+# The performance of our linear regression model with this simple feature
+# engineering is a bit better than using the original ordinal time features but
+# worse than using the one-hot encoded time features. We will further analyze
+# possible reasons for this disappointing outcome at the end of this notebook.
+#
+# Periodic spline features
+# ------------------------
+#
+# We can try an alternative encoding of the periodic time-related features
+# using spline transformations with a large-enough number of knots, and as a
+# result a larger number of expanded features:
+from sklearn.preprocessing import SplineTransformer
+
+
+def periodic_spline_transformer(period, n_knots=None, degree=3):
+    if n_knots is None:
+        n_knots = period
+    return SplineTransformer(
+        degree=degree,
+        n_knots=n_knots,
+        knots=np.linspace(0, period, n_knots).reshape(n_knots, 1),
+        extrapolation="periodic",
+    )
+
+
+# %%
+#
+# Again, let us visualize the effect of this feature expansion on some
+# synthetic hour data with a bit of extrapolation beyond hour=23:
+hour_df = pd.DataFrame(
+    np.linspace(0, 26, 1000).reshape(-1, 1),
+    columns=["hour"],
+)
+splines = periodic_spline_transformer(24, n_knots=12).fit_transform(hour_df)
+splines_df = pd.DataFrame(
+    splines,
+    columns=[f"spline_{i}" for i in range(splines.shape[1])],
+)
+pd.concat([hour_df, splines_df], axis="columns").plot(x="hour")
+_ = plt.title("Periodic spline-based encoding for the 'hour' feature")
+
+
+# %%
+# Thanks for the use of the `extrapolation="periodic"` parameter, we observe
+# that the feature encoding stays smooth when extrapolating beyond midnight.
+#
+# We can now build a predictive pipeline using this alternative periodic
+# feature engineering strategy.
+#
+# For the "hours" columns we use only 12 knots for a period of 24 hours to
+# avoid over-representing this feature compared to months (12 natural knots)
+# and weekday (7 natural knots).
+cyclic_spline_transformer = ColumnTransformer(
+    transformers=[
+        ("categorical", one_hot_encoder, categorical_columns),
+        ("cyclic_month", periodic_spline_transformer(12), ["month"]),
+        ("cyclic_weekday", periodic_spline_transformer(7), ["weekday"]),
+        ("cyclic_hour", periodic_spline_transformer(24, n_knots=12), ["hour"]),
+    ],
+    remainder="passthrough",
+)
+cyclic_spline_linear_pipeline = make_pipeline(
+    cyclic_spline_transformer,
+    RidgeCV(alphas=alphas),
+)
+evaluate(cyclic_spline_linear_pipeline, X, y, cv=ts_cv)
+
+
+# %%
+# Spline features make it possible for the linear model to successfully
+# leverage the periodic time-related features and reduce the error from ~14% to
+# ~10% of the maximum demand, which is similar to what we observed with the
+# one-hot encoded features.
```

Comment:
Could be interesting to also compare the number of coefficients=degree of freedoms.

Reply:
I added a new paragraph + code later in the notebook, once each pipeline is fit to compare the resulting number of features.